### ISSUE TITLE
`torch.scatter_add_` appears to be (about 5x) slower than cuda backend

### ISSUE THREAD

**Issue by fcharras** [intel/intel-extension-for-pytorch#426]
### Describe the bug

The following benchmark on Max Series with xpu backend

```ipex
import intel_extension_for_pytorch as ipex

import torch
device = "xpu"
n_samples = 50_000_000
dim = 10
n_scatters = 100
data = torch.rand(n_samples, dim, dtype=torch.float32, device=device)
group_idx = torch.randint(0, n_scatters, (n_samples,1), dtype=torch.int64, device=device)
groupby_sum = torch.zeros(n_scatters, dim, dtype=torch.float32, device=device)
%timeit groupby_sum.scatter_add(dim=0, index=group_idx.expand(-1, dim), src=data)[0, 0].cpu().item()
```

prints

```
79.5 ms ± 14.9 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

The same on nvidia A100 with cuda backend

```python
import torch
device = "cuda"
n_samples = 50_000_000
dim = 10
n_scatters = 100
data = torch.rand(n_samples, dim, dtype=torch.float32, device=device)
group_idx = torch.randint(0, n_scatters, (n_samples,1), dtype=torch.int64, device=device)
groupby_sum = torch.zeros(n_scatters, dim, dtype=torch.float32, device=device)
%timeit groupby_sum.scatter_add(dim=0, index=group_idx.expand(-1, dim), src=data)[0,0].cpu().item()
```

prints

```
13.7 ms ± 902 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

NB: the `.cpu().item()` is necessary to ensure the benchmark is not affected by async compute.

Hunch about how to explain the difference: the values for `n_samples`, `dim` and `n_scatters` chosen here, create a high probability of collision on atomic add operations under the hood. There are implementation tricks that can mitigate the performance hit coming from the conflicts. From experience from similar kernel programming work I've been involved with, I'd assume those tricks could be the difference between xpu / cuda backends here (and not the gpu specs)

Context / impact: I'm benchmarking a portable implementation of a kmeans algorithm using `pytorch`, [`torch.expand` + `torch.scatter_add_`](https://github.com/soda-inria/sklearn-pytorch-engine/blob/main/sklearn_pytorch_engine/kmeans/engine.py#L458) is necessary for the centroid update step. With `xpu` backend it becomes a bottleneck, but it's not a bottleneck with `cuda` backend.

KMeans with pytorch seems that it might be important, I see a nice speedup over optimized CPU implementations, and it's easy to install/distribute, and has portability by design in pytorch, so it's a really interesting backend choice (on the minus side it seems to be about 3x slower than low-level optimized gpu implementations that can avoid the IO bottleneck when writing intermediate arrays in memory by fusing kernels together ~ but still interesting compared to CPU)

### Versions

I used the conda install of the most recent relase.

```
Collecting environment information...
PyTorch version: 2.0.1a0+cxx11.abi
PyTorch CXX11 ABI: Yes
IPEX version: 2.0.110+xpu
IPEX commit: ba7f6c127
Build type: Release

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: N/A
IGC version: N/A
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jul 20 2023, 00:27:19) [GCC 13.1.0] (64-bit runtime)
Python platform: Linux-5.15.0-83-generic-x86_64-with-glibc2.35
Is XPU available: True
DPCPP runtime version: N/A
MKL version: N/A
GPU models and configuration: 
[0] _DeviceProperties(name='Intel(R) Data Center GPU Max 1100', platform_name='Intel(R) Level-Zero', dev_type='gpu, support_fp64=1, total_memory=49152MB, max_compute_units=448, gpu_eu_count=448)
Intel OpenCL ICD version: 23.22.26516.25-682~22.04
Level Zero version: 1.3.26516.25-682~22.04

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      52 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             224
On-line CPU(s) list:                0-223
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Platinum 8480+
CPU family:                         6
Model:                              143
Thread(s) per core:                 2
Core(s) per socket:                 56
Socket(s):                          2
Stepping:                           8
CPU max MHz:                        3800.0000
CPU min MHz:                        800.0000
BogoMIPS:                           4000.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          5.3 MiB (112 instances)
L1i cache:                          3.5 MiB (112 instances)
L2 cache:                           224 MiB (112 instances)
L3 cache:                           210 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-55,112-167
NUMA node1 CPU(s):                  56-111,168-223
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] intel-extension-for-pytorch==2.0.110+xpu
[pip3] numpy==1.24.3
[pip3] sklearn-pytorch-engine==0.1.0.dev0
[pip3] torch==2.0.1a0+cxx11.abi
[conda] intel-extension-for-pytorch 2.0.110             py310_xpu_0    intel
[conda] mkl                       2023.2.0            intel_49495    intel
[conda] mkl-dpcpp                 2023.2.0            intel_49495    intel
[conda] mkl-service               2.4.0           py310hae59892_35    intel
[conda] mkl_fft                   1.3.6           py310h173b8ae_56    intel
[conda] mkl_random                1.2.2           py310h1595b48_76    intel
[conda] mkl_umath                 0.1.1           py310hd987cd3_86    intel
[conda] numpy                     1.24.3          py310hed7eef7_0    intel
[conda] numpy-base                1.24.3          py310he88ecf9_0    intel
[conda] pytorch                   2.0.1               py310_xpu_0    intel
[conda] sklearn-pytorch-engine    0.1.0.dev0               pypi_0    pypi
```

**Comment by jingxu10** [comment]
@tye1 @gujinghui
